source/time_stepping/lapusta_time_step.cc

```diff
@@ -0,0 +1,183 @@
+                                                                                        in.position[q],
+                                                                                        delta_x,
+                                                                                        in.pressure[q]);
+
+                min_state_weakening_time_step = std::min (min_state_weakening_time_step,
+                                                          delta_theta_max_and_critical_slip_distance.first
+                                                          * delta_theta_max_and_critical_slip_distance.second / local_velocity);
+
+                // state healing time step is: Deltat_h = 0.2 * theta.
+                min_healing_time_step = std::min (min_healing_time_step,
+                                                  viscoplastic.compute_min_healing_time_step(in.composition[q]));
+
+                // the maximum local velocity needed for the displacement time step
+                max_local_velocity = std::max (max_local_velocity,
+                                               local_velocity);
+
+                // the viscoelastoplastic relaxation time step using the relaxation time scale: f_max * viscoplastic viscosity / shear modulus
+                // with f_max = 0.2 in Herrendörfer et al. 2018
+                // to capture the increasing slip rate in case of a purely rate-dependent friction, i.e. if b = 0
+                min_vep_relaxation_time_step = std::min (min_vep_relaxation_time_step,
+                                                         0.2 * out.viscosities[q]
+                                                         / viscoplastic.get_elastic_shear_modulus(in.composition[0]));
+              }
+
+            // minimum displacement time step: Delta t_d = Delta d_max * min(|Delta x/v_x|,|Delta x/v_y|),
+            // with Delta d_max = 1e-3 in Herrendörfer et al. 2018
+            // here, the term  min(|Delta x/v_x|,|Delta x/v_y|) is simplified to min Delta x / max_local_velocity
+            min_displacement_time_step = std::min (min_displacement_time_step,
+                                                   1.e-3 * cell->minimum_vertex_distance() / max_local_velocity);
+
+          }
+
+      double min_lapusta_timestep = std::numeric_limits<double>::max();
+
+      // take the minimum of the four criteria
+      // TODO: is there a more elegant way to take the minimum of four values?
+      min_lapusta_timestep = std::min (min_state_weakening_time_step,
+                                       std::min (min_healing_time_step,
+                                                 std::min (min_displacement_time_step,
+                                                           min_vep_relaxation_time_step)));
+
```

Comment:
It seems you don't communicate the minimum timestep between the processes. You need something like:
```
const double min_global_lapusta_timestep
 = Utilities::MPI::min (min_lapusta_timestep, this->get_mpi_communicator());
```